erpnext/stock/doctype/stock_entry/test_stock_entry.py

```diff
@@ -73,6 +73,174 @@ def test_fifo(self):
 
 		frappe.db.set_default("allow_negative_stock", 0)
 
+	def test_batch_wise_valuation(self):
+		item_code = "_Test Batch Wise Valuation Item"
+		warehouse = "_Test Warehouse - _TC"
+
+		for d in frappe.get_all("Stock Entry", filters={'item_code': item_code, 'docstatus': 1}):
+			frappe.get_doc("Stock Entry", d.name).cancel()
+
+		def assert_sle(sle, values):
+			for k, v in iteritems(values):
+				self.assertEqual(sle.get(k), v)
+
+		# In 1
+		in1_1 = make_stock_entry(item_code=item_code, target=warehouse, qty=10, basic_rate=100)
+		batch1 = in1_1.items[0].batch_no
+		assert_sle(get_sle(item_code=item_code, warehouse=warehouse, batch_no=batch1,
+			voucher_type=in1_1.doctype, voucher_no=in1_1.name)[0], {
+			'stock_value_difference': 1000, 'batch_stock_value': 1000, 'stock_value': 1000,
+			'batch_valuation_rate': 100, 'valuation_rate': 100,
+			'batch_qty_after_transaction': 10, 'qty_after_transaction': 10
+		})
+
+		in1_2 = make_stock_entry(item_code=item_code, target=warehouse, qty=10, basic_rate=80)
```

Comment:
break these into separate tests? also please add comments for each of these tests. So that reviewing is easy.

Reply:
Can't be broken, they're all dependent on each other. Comments I can do

Reply:
The test is quite self explanatory. For a detailed understanding only the spreadsheet at https://docs.google.com/spreadsheets/d/10GOG70ZuKPm4pjF8ZpvjniNHmo05i6Gs6jJedhaXKso/edit?usp=sharing will help. Added a link in comments